{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import spacy\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/10/s1vq2hq12czcjxjzsn6z4qf40000gn/T/ipykernel_88693/3701423186.py:2: DtypeWarning: Columns (69,70,71,75,76) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv('metadata.csv')\n"
     ]
    }
   ],
   "source": [
    "# Read in Metadata and handle IDs\n",
    "df = pd.read_csv('metadata.csv')\n",
    "df[['Taxon Object ID', 'IMG Submission ID']] = df[['Taxon Object ID', 'IMG Submission ID']].fillna(0).astype(int)\n",
    "\n",
    "# Handle Locations\n",
    "def extract_location_from_text(input_text):\n",
    "    doc = nlp(input_text)\n",
    "    location_names = [entity.text for entity in doc.ents if entity.label_ == 'GPE']\n",
    "\n",
    "    return location_names\n",
    "\n",
    "# Find all rows with null values in 'Geographic Location' that have non-null values in 'Latitude' and 'Longitude'\n",
    "null_locations = list(df[df['Geographic Location'].isnull()].index)\n",
    "\n",
    "# Use spacy nlp to extract locations from project names and impute, where possible\n",
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "for idx in list(df['Study Name (Proposal Name)'].iloc[null_locations].index):\n",
    "    extracted_loc = extract_location_from_text(str(df['Study Name (Proposal Name)'].iloc[idx]))\n",
    "\n",
    "    if extracted_loc and extracted_loc != ['nan']:\n",
    "        extracted_loc = ' '.join(extracted_loc)\n",
    "        df.loc[df.index[idx], 'Geographic Location'] = extracted_loc\n",
    "\n",
    "    # Check if there are 'various locations' listed and impute if so\n",
    "    elif not extracted_loc:\n",
    "        if re.search('various locations', str(df['Study Name (Proposal Name)'].iloc[idx])):\n",
    "            df.loc[df.index[idx], 'Geographic Location'] = 'various locations'\n",
    "\n",
    "# Read in Portals\n",
    "portals_df = pd.read_csv('portals.csv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study Name (Proposal Name)</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Taxon Object ID</th>\n",
       "      <th>IMG Submission ID</th>\n",
       "      <th>\\nGOLD IDs in IMG Database</th>\n",
       "      <th>GOLD Analysis Project Type</th>\n",
       "      <th>Submission Type</th>\n",
       "      <th>JGI Analysis Project Type</th>\n",
       "      <th>SRA ID</th>\n",
       "      <th>SRA Run</th>\n",
       "      <th>...</th>\n",
       "      <th>Proportal Isolation</th>\n",
       "      <th>Proportal Ocean</th>\n",
       "      <th>Proportal Station</th>\n",
       "      <th>Proportal WOA Nitrate</th>\n",
       "      <th>Proportal WOA Salinity</th>\n",
       "      <th>Proportal WOA Temperature</th>\n",
       "      <th>GenBank Anomalous Assembly</th>\n",
       "      <th>NCBI Assembly Accession</th>\n",
       "      <th>portal_name</th>\n",
       "      <th>downloaded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Marine microbial communities from different lo...</td>\n",
       "      <td>Marine microbial communities from Red Sea, Sau...</td>\n",
       "      <td>3300038838</td>\n",
       "      <td>219915</td>\n",
       "      <td>\\nStudy ID: Gs0132945  Project ID: Gp0274035  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>SRX1097578</td>\n",
       "      <td>SRR2102997</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMG_3300038838</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Agricultural soil microbial communities from A...</td>\n",
       "      <td>Agricultural soil microbial communities from s...</td>\n",
       "      <td>3300064906</td>\n",
       "      <td>315709</td>\n",
       "      <td>\\nStudy ID: Gs0161729  Project ID: Gp0758376  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMG_3300064906</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Freshwater microbial mat bacterial communities...</td>\n",
       "      <td>Freshwater microbial mat bacterial communities...</td>\n",
       "      <td>3300015360</td>\n",
       "      <td>123175</td>\n",
       "      <td>\\nStudy ID: Gs0127369  Project ID: Gp0191361  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>SRX3539175, SRX3539174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OliLak19BULKMAT1_FD</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Soil microbial communities from Los Alamos Nat...</td>\n",
       "      <td>Soil microbial communities from Los Alamos Nat...</td>\n",
       "      <td>3300053171</td>\n",
       "      <td>270981</td>\n",
       "      <td>\\nStudy ID: Gs0153999  Project ID: Gp0565207  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S17hydmetagenome_6_FD</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tropical forest soil microbial communities fro...</td>\n",
       "      <td>Tropical forest soil microbial communities fro...</td>\n",
       "      <td>3300000729</td>\n",
       "      <td>11900</td>\n",
       "      <td>\\nStudy ID: Gs0075432  Project ID: Gp0054556  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Reanalysis</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>SRX4340826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LuqExpMetageno35_FD</td>\n",
       "      <td>no</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Study Name (Proposal Name)  \\\n",
       "0  Marine microbial communities from different lo...   \n",
       "1  Agricultural soil microbial communities from A...   \n",
       "2  Freshwater microbial mat bacterial communities...   \n",
       "3  Soil microbial communities from Los Alamos Nat...   \n",
       "4  Tropical forest soil microbial communities fro...   \n",
       "\n",
       "                                         Sample Name  Taxon Object ID  \\\n",
       "0  Marine microbial communities from Red Sea, Sau...       3300038838   \n",
       "1  Agricultural soil microbial communities from s...       3300064906   \n",
       "2  Freshwater microbial mat bacterial communities...       3300015360   \n",
       "3  Soil microbial communities from Los Alamos Nat...       3300053171   \n",
       "4  Tropical forest soil microbial communities fro...       3300000729   \n",
       "\n",
       "   IMG Submission ID                         \\nGOLD IDs in IMG Database  \\\n",
       "0             219915  \\nStudy ID: Gs0132945  Project ID: Gp0274035  ...   \n",
       "1             315709  \\nStudy ID: Gs0161729  Project ID: Gp0758376  ...   \n",
       "2             123175  \\nStudy ID: Gs0127369  Project ID: Gp0191361  ...   \n",
       "3             270981  \\nStudy ID: Gs0153999  Project ID: Gp0565207  ...   \n",
       "4              11900  \\nStudy ID: Gs0075432  Project ID: Gp0054556  ...   \n",
       "\n",
       "  GOLD Analysis Project Type Submission Type JGI Analysis Project Type  \\\n",
       "0        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "1        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "2        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "3        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "4        Metagenome Analysis      Reanalysis       Metagenome Analysis   \n",
       "\n",
       "                   SRA ID     SRA Run  ... Proportal Isolation  \\\n",
       "0              SRX1097578  SRR2102997  ...                 NaN   \n",
       "1                     NaN         NaN  ...                 NaN   \n",
       "2  SRX3539175, SRX3539174         NaN  ...                 NaN   \n",
       "3                     NaN         NaN  ...                 NaN   \n",
       "4              SRX4340826         NaN  ...                 NaN   \n",
       "\n",
       "  Proportal Ocean Proportal Station Proportal WOA Nitrate  \\\n",
       "0             NaN               NaN                   NaN   \n",
       "1             NaN               NaN                   NaN   \n",
       "2             NaN               NaN                   NaN   \n",
       "3             NaN               NaN                   NaN   \n",
       "4             NaN               NaN                   NaN   \n",
       "\n",
       "  Proportal WOA Salinity Proportal WOA Temperature GenBank Anomalous Assembly  \\\n",
       "0                    NaN                       NaN                        NaN   \n",
       "1                    NaN                       NaN                        NaN   \n",
       "2                    NaN                       NaN                        NaN   \n",
       "3                    NaN                       NaN                        NaN   \n",
       "4                    NaN                       NaN                        NaN   \n",
       "\n",
       "  NCBI Assembly Accession            portal_name downloaded  \n",
       "0                     NaN         IMG_3300038838         no  \n",
       "1                     NaN         IMG_3300064906         no  \n",
       "2                     NaN    OliLak19BULKMAT1_FD         no  \n",
       "3                     NaN  S17hydmetagenome_6_FD         no  \n",
       "4                     NaN    LuqExpMetageno35_FD         no  \n",
       "\n",
       "[5 rows x 79 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join Metadata with Portal Names\n",
    "df = df.join(portals_df.set_index('taxon_oid'), on='Taxon Object ID')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows without Study Names\n",
    "df = df[-df['Study Name (Proposal Name)'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Study ID'] = None\n",
    "df['Project ID'] = None\n",
    "df['Analysis ID'] = None\n",
    "\n",
    "for mult_ids in list(df['\\nGOLD IDs in IMG Database']):\n",
    "    id_list = mult_ids.strip('\\n').split('\\xa0\\xa0')\n",
    "\n",
    "    for id_ in id_list:\n",
    "        if id_ != '':\n",
    "            id_type = id_.split(': ')[0]\n",
    "            id_num = id_.split(': ')[1]\n",
    "            df.loc[df['\\nGOLD IDs in IMG Database'] == mult_ids, id_type] = id_num\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add contigs.fna files to metadata\n",
    "### CHANGE THIS PATH\n",
    "downloaded = os.listdir('/Users/kaenurge/Desktop/G2 Lab/crisprdata_test/')\n",
    "\n",
    "### COMMENT THIS OUT IF RUNNING ANOTHER BATCH\n",
    "df['files downloaded'] = None\n",
    "\n",
    "for file in downloaded:\n",
    "    project = file.split('__')[0]\n",
    "    if project in list(df['portal_name']):\n",
    "        idx = df.loc[df['portal_name'] == project].index\n",
    "\n",
    "        # Handle cases where there are multiple projects with the same name\n",
    "        if len(idx) > 1:\n",
    "            id_cols = ['Study ID', 'Project ID', 'Analysis ID']\n",
    "            file_name_extended = file.split('__')[1]\n",
    "            file_name = file_name_extended.split('_')[0]\n",
    "            condition_mask = df[id_cols].apply(lambda col: col.str.contains(file_name)).any(axis=1) #if id is not found, make idx = idx[0] and make another column to note it\n",
    "            idx = df[condition_mask].index\n",
    "        \n",
    "        idx = int(idx[0])\n",
    "\n",
    "        # Append file name, handle cases if more than 1 file is downloaded for the same project\n",
    "        if df.at[idx, 'files downloaded'] == None:\n",
    "            df.at[idx, 'files downloaded'] = file\n",
    "        else:\n",
    "            df.at[idx, 'files downloaded'] += f', {file}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Ga0455849'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# a = '003ER18SCSDNA_FD__Ga0455849_contigs_ftd.fna'\n",
    "# b = a.split('__')[1]\n",
    "# b.split('_')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([], dtype='int64')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# id_cols = ['Study ID', 'Project ID', 'Analysis ID']\n",
    "# condition_mask = df[id_cols].apply(lambda col: col.str.contains('003ER18SCSDNA_FD')).any(axis=1)\n",
    "# G2 Lab/crisprdata_test/003ER18SCSDNA_FD__Ga0455849_contigs_ftd.fna\n",
    "# df[condition_mask].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study Name (Proposal Name)</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Taxon Object ID</th>\n",
       "      <th>IMG Submission ID</th>\n",
       "      <th>\\nGOLD IDs in IMG Database</th>\n",
       "      <th>GOLD Analysis Project Type</th>\n",
       "      <th>Submission Type</th>\n",
       "      <th>JGI Analysis Project Type</th>\n",
       "      <th>SRA ID</th>\n",
       "      <th>SRA Run</th>\n",
       "      <th>...</th>\n",
       "      <th>Proportal WOA Salinity</th>\n",
       "      <th>Proportal WOA Temperature</th>\n",
       "      <th>GenBank Anomalous Assembly</th>\n",
       "      <th>NCBI Assembly Accession</th>\n",
       "      <th>portal_name</th>\n",
       "      <th>downloaded</th>\n",
       "      <th>Study ID</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Analysis ID</th>\n",
       "      <th>files downloaded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Marine microbial communities from different lo...</td>\n",
       "      <td>Marine microbial communities from Red Sea, Sau...</td>\n",
       "      <td>3300038838</td>\n",
       "      <td>219915</td>\n",
       "      <td>\\nStudy ID: Gs0132945  Project ID: Gp0274035  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>SRX1097578</td>\n",
       "      <td>SRR2102997</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMG_3300038838</td>\n",
       "      <td>no</td>\n",
       "      <td>Gs0132945</td>\n",
       "      <td>Gp0274035</td>\n",
       "      <td>Ga0237899</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Agricultural soil microbial communities from A...</td>\n",
       "      <td>Agricultural soil microbial communities from s...</td>\n",
       "      <td>3300064906</td>\n",
       "      <td>315709</td>\n",
       "      <td>\\nStudy ID: Gs0161729  Project ID: Gp0758376  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMG_3300064906</td>\n",
       "      <td>no</td>\n",
       "      <td>Gs0161729</td>\n",
       "      <td>Gp0758376</td>\n",
       "      <td>Ga0613092</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Freshwater microbial mat bacterial communities...</td>\n",
       "      <td>Freshwater microbial mat bacterial communities...</td>\n",
       "      <td>3300015360</td>\n",
       "      <td>123175</td>\n",
       "      <td>\\nStudy ID: Gs0127369  Project ID: Gp0191361  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>SRX3539175, SRX3539174</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OliLak19BULKMAT1_FD</td>\n",
       "      <td>no</td>\n",
       "      <td>Gs0127369</td>\n",
       "      <td>Gp0191361</td>\n",
       "      <td>Ga0163144</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Soil microbial communities from Los Alamos Nat...</td>\n",
       "      <td>Soil microbial communities from Los Alamos Nat...</td>\n",
       "      <td>3300053171</td>\n",
       "      <td>270981</td>\n",
       "      <td>\\nStudy ID: Gs0153999  Project ID: Gp0565207  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S17hydmetagenome_6_FD</td>\n",
       "      <td>no</td>\n",
       "      <td>Gs0153999</td>\n",
       "      <td>Gp0565207</td>\n",
       "      <td>Ga0494674</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tropical forest soil microbial communities fro...</td>\n",
       "      <td>Tropical forest soil microbial communities fro...</td>\n",
       "      <td>3300000729</td>\n",
       "      <td>11900</td>\n",
       "      <td>\\nStudy ID: Gs0075432  Project ID: Gp0054556  ...</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>Reanalysis</td>\n",
       "      <td>Metagenome Analysis</td>\n",
       "      <td>SRX4340826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LuqExpMetageno35_FD</td>\n",
       "      <td>no</td>\n",
       "      <td>Gs0075432</td>\n",
       "      <td>Gp0054556</td>\n",
       "      <td>Ga0001958</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 83 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Study Name (Proposal Name)  \\\n",
       "0  Marine microbial communities from different lo...   \n",
       "1  Agricultural soil microbial communities from A...   \n",
       "2  Freshwater microbial mat bacterial communities...   \n",
       "3  Soil microbial communities from Los Alamos Nat...   \n",
       "4  Tropical forest soil microbial communities fro...   \n",
       "\n",
       "                                         Sample Name  Taxon Object ID  \\\n",
       "0  Marine microbial communities from Red Sea, Sau...       3300038838   \n",
       "1  Agricultural soil microbial communities from s...       3300064906   \n",
       "2  Freshwater microbial mat bacterial communities...       3300015360   \n",
       "3  Soil microbial communities from Los Alamos Nat...       3300053171   \n",
       "4  Tropical forest soil microbial communities fro...       3300000729   \n",
       "\n",
       "   IMG Submission ID                         \\nGOLD IDs in IMG Database  \\\n",
       "0             219915  \\nStudy ID: Gs0132945  Project ID: Gp0274035  ...   \n",
       "1             315709  \\nStudy ID: Gs0161729  Project ID: Gp0758376  ...   \n",
       "2             123175  \\nStudy ID: Gs0127369  Project ID: Gp0191361  ...   \n",
       "3             270981  \\nStudy ID: Gs0153999  Project ID: Gp0565207  ...   \n",
       "4              11900  \\nStudy ID: Gs0075432  Project ID: Gp0054556  ...   \n",
       "\n",
       "  GOLD Analysis Project Type Submission Type JGI Analysis Project Type  \\\n",
       "0        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "1        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "2        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "3        Metagenome Analysis         Primary       Metagenome Analysis   \n",
       "4        Metagenome Analysis      Reanalysis       Metagenome Analysis   \n",
       "\n",
       "                   SRA ID     SRA Run  ... Proportal WOA Salinity  \\\n",
       "0              SRX1097578  SRR2102997  ...                    NaN   \n",
       "1                     NaN         NaN  ...                    NaN   \n",
       "2  SRX3539175, SRX3539174         NaN  ...                    NaN   \n",
       "3                     NaN         NaN  ...                    NaN   \n",
       "4              SRX4340826         NaN  ...                    NaN   \n",
       "\n",
       "  Proportal WOA Temperature GenBank Anomalous Assembly  \\\n",
       "0                       NaN                        NaN   \n",
       "1                       NaN                        NaN   \n",
       "2                       NaN                        NaN   \n",
       "3                       NaN                        NaN   \n",
       "4                       NaN                        NaN   \n",
       "\n",
       "  NCBI Assembly Accession            portal_name downloaded   Study ID  \\\n",
       "0                     NaN         IMG_3300038838         no  Gs0132945   \n",
       "1                     NaN         IMG_3300064906         no  Gs0161729   \n",
       "2                     NaN    OliLak19BULKMAT1_FD         no  Gs0127369   \n",
       "3                     NaN  S17hydmetagenome_6_FD         no  Gs0153999   \n",
       "4                     NaN    LuqExpMetageno35_FD         no  Gs0075432   \n",
       "\n",
       "  Project ID Analysis ID files downloaded  \n",
       "0  Gp0274035   Ga0237899             None  \n",
       "1  Gp0758376   Ga0613092             None  \n",
       "2  Gp0191361   Ga0163144             None  \n",
       "3  Gp0565207   Ga0494674             None  \n",
       "4  Gp0054556   Ga0001958             None  \n",
       "\n",
       "[5 rows x 83 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('final_metadata.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Portal Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>CRISPR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Portal Name, ID, CRISPR]\n",
       "Index: []"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add contigs.fna files to metadata\n",
    "### CHANGE THIS PATH\n",
    "directory_path = '/Users/kaenurge/Desktop/G2 Lab/crisprdata copy 2/'\n",
    "cctyped = os.listdir(directory_path)\n",
    "\n",
    "portals = pd.read_csv('portals.csv', sep='\\t')\n",
    "portals.rename(columns = {'taxon_oid': 'Taxon Object ID'}, inplace = True)\n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['Portal Name'] = None\n",
    "df['ID'] = None\n",
    "df['CRISPR'] = None\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "for directory in cctyped:\n",
    "    file_path = directory_path + str(directory)\n",
    "\n",
    "    # Loop through directories created by cctyper\n",
    "    if os.path.isdir(file_path):\n",
    "        project = directory.split('__')[0]\n",
    "        id_ = directory.split('__')[1]\n",
    "        cctype_output = os.listdir(file_path)\n",
    "\n",
    "        # Check if CRISPRs were found\n",
    "        if 'crisprs_all.tab' in cctype_output:\n",
    "            file_path += '/crisprs_all.tab'\n",
    "            all_crisprs = pd.read_csv(file_path, sep = '\\t')\n",
    "            temp_df = pd.DataFrame(all_crisprs)\n",
    "            temp_df['Portal Name'] = project\n",
    "            temp_df['ID'] = id_\n",
    "            \n",
    "            df = pd.concat([df, temp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Portal Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>CRISPR</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Consensus_repeat</th>\n",
       "      <th>N_repeats</th>\n",
       "      <th>Repeat_len</th>\n",
       "      <th>Spacer_len_avg</th>\n",
       "      <th>Repeat_identity</th>\n",
       "      <th>Spacer_identity</th>\n",
       "      <th>Spacer_len_sem</th>\n",
       "      <th>Trusted</th>\n",
       "      <th>Prediction</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>Subtype_probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>182ER18SCSDNA_FD</td>\n",
       "      <td>Ga0454634</td>\n",
       "      <td>Ga0454634_0031049_1</td>\n",
       "      <td>Ga0454634_0031049</td>\n",
       "      <td>815.0</td>\n",
       "      <td>979.0</td>\n",
       "      <td>GTCGCACCCCACGCGGGTGCGTGAATTGAAAC</td>\n",
       "      <td>3.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>34.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>True</td>\n",
       "      <td>I-C</td>\n",
       "      <td>I-C</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>182ER18SCSDNA_FD</td>\n",
       "      <td>Ga0454634</td>\n",
       "      <td>Ga0454634_0053810_2</td>\n",
       "      <td>Ga0454634_0053810</td>\n",
       "      <td>105.0</td>\n",
       "      <td>804.0</td>\n",
       "      <td>GGTTCCCCCGCCTGCGCGGGGATGGCCCC</td>\n",
       "      <td>12.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>89.5</td>\n",
       "      <td>45.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "      <td>I-E</td>\n",
       "      <td>I-E</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>182ER18SCSDNA_FD</td>\n",
       "      <td>Ga0454634</td>\n",
       "      <td>Ga0454634_0080066_3</td>\n",
       "      <td>Ga0454634_0080066</td>\n",
       "      <td>1.0</td>\n",
       "      <td>242.0</td>\n",
       "      <td>GCGGCGAGGACCGCGCTCGCCGACG</td>\n",
       "      <td>4.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>47.3</td>\n",
       "      <td>79.3</td>\n",
       "      <td>52.7</td>\n",
       "      <td>1.5</td>\n",
       "      <td>True</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>I-C</td>\n",
       "      <td>0.437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>182ER18SCSDNA_FD</td>\n",
       "      <td>Ga0454634</td>\n",
       "      <td>Ga0454634_0126594_4</td>\n",
       "      <td>Ga0454634_0126594</td>\n",
       "      <td>264.0</td>\n",
       "      <td>518.0</td>\n",
       "      <td>ATCTCCGTCGACGTTCGTCGGCGGCCCCATTGAAGC</td>\n",
       "      <td>4.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>47.2</td>\n",
       "      <td>0.6</td>\n",
       "      <td>True</td>\n",
       "      <td>I-G</td>\n",
       "      <td>I-G</td>\n",
       "      <td>0.999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Portal Name          ID               CRISPR             Contig  \\\n",
       "0  182ER18SCSDNA_FD  Ga0454634   Ga0454634_0031049_1  Ga0454634_0031049   \n",
       "1  182ER18SCSDNA_FD  Ga0454634   Ga0454634_0053810_2  Ga0454634_0053810   \n",
       "2  182ER18SCSDNA_FD  Ga0454634   Ga0454634_0080066_3  Ga0454634_0080066   \n",
       "3  182ER18SCSDNA_FD  Ga0454634   Ga0454634_0126594_4  Ga0454634_0126594   \n",
       "\n",
       "   Start    End                      Consensus_repeat  N_repeats  Repeat_len  \\\n",
       "0  815.0  979.0      GTCGCACCCCACGCGGGTGCGTGAATTGAAAC        3.0        32.0   \n",
       "1  105.0  804.0         GGTTCCCCCGCCTGCGCGGGGATGGCCCC       12.0        29.0   \n",
       "2    1.0  242.0             GCGGCGAGGACCGCGCTCGCCGACG        4.0        25.0   \n",
       "3  264.0  518.0  ATCTCCGTCGACGTTCGTCGGCGGCCCCATTGAAGC        4.0        36.0   \n",
       "\n",
       "   Spacer_len_avg  Repeat_identity  Spacer_identity  Spacer_len_sem Trusted  \\\n",
       "0            34.5            100.0             50.0             0.5    True   \n",
       "1            32.0             89.5             45.8             0.0    True   \n",
       "2            47.3             79.3             52.7             1.5    True   \n",
       "3            37.0            100.0             47.2             0.6    True   \n",
       "\n",
       "  Prediction Subtype  Subtype_probability  \n",
       "0        I-C     I-C                1.000  \n",
       "1        I-E     I-E                1.000  \n",
       "2    Unknown     I-C                0.437  \n",
       "3        I-G     I-G                0.999  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
